webdav_storage/src/WebDav.php

```diff
@@ -0,0 +1,369 @@
+	 */
+	protected function hasItems(string $uri): bool
+	{
+		$dom               = new \DOMDocument('1.0', 'UTF-8');
+		$dom->formatOutput = true;
+		$root              = $dom->createElementNS('DAV:', 'd:propfind');
+		$prop              = $dom->createElement('d:allprop');
+
+		$dom->appendChild($root)->appendChild($prop);
+
+		$opts = [
+			HTTPClientOptions::AUTH    => $this->authOptions,
+			HTTPClientOptions::HEADERS => ['Depth' => 1, 'Prefer' => 'return-minimal', 'Content-Type' => 'application/xml'],
+			HTTPClientOptions::BODY    => $dom->saveXML(),
+		];
+
+		$response = $this->client->request('propfind', $uri, $opts);
+
+		$responseDoc = new \DOMDocument();
+		$responseDoc->loadXML($response->getBody());
+		$responseDoc->formatOutput = true;
+
+		$xpath = new \DOMXPath($responseDoc);
+		$xpath->registerNamespace('d', 'DAV');
+		$result = $xpath->query('//d:multistatus/d:response');
+
+		// returns at least its own directory, so >1
+		return $result !== false && count($result) > 1;
+	}
+
+	/**
+	 * Creates a DAV-collection (= folder) for the given uri
+	 *
+	 * @param string $uri The uri for creating a DAV-collection
+	 *
+	 * @return bool true in case the creation was successful (not immutable!)
+	 */
+	protected function mkcol(string $uri): bool
+	{
+		return $this->client->request('mkcol', $uri, [HTTPClientOptions::AUTH => $this->authOptions])
+							->getReturnCode() == 200;
+	}
+
+	/**
+	 * Checks if the given path exists and if not creates it
+	 *
+	 * @param string $fullPath the full path (the folder structure after the hostname)
+	 */
+	protected function checkAndCreatePath(string $fullPath): void
+	{
+		$finalUrl = $this->url . '/' . trim($fullPath, '/');
+
+		if ($this->exists($finalUrl)) {
+			return;
+		}
+
+		$pathParts = explode('/', trim($fullPath, '/'));
+		$path      = '';
+
+		foreach ($pathParts as $part) {
+			$path .= '/' . $part;
+			$partUrl = $this->url . $path;
+			if (!$this->exists($partUrl)) {
+				$this->mkcol($partUrl);
+			}
+		}
+	}
+
+	/**
+	 * Checks recursively, if paths are empty and deletes them
+	 *
+	 * @param string $fullPath the full path (the folder structure after the hostname)
+	 *
+	 * @throws StorageException In case a directory cannot get deleted
+	 */
+	protected function checkAndDeletePath(string $fullPath): void
+	{
+		$pathParts = explode('/', trim($fullPath, '/'));
+		$partURL   = '/' . implode('/', $pathParts);
+
+		foreach ($pathParts as $pathPart) {
+			$checkUrl = $this->url . $partURL;
+			if (!empty($partURL) && !$this->hasItems($checkUrl)) {
+				$response = $this->client->request('delete', $checkUrl, [HTTPClientOptions::AUTH => $this->authOptions]);
+
+				if (!$response->isSuccess()) {
+					if ($response->getReturnCode() == "404") {
+						$this->logger->warning('Directory already deleted.', ['uri' => $checkUrl]);
+					} else {
+						throw new StorageException(sprintf('Unpredicted error for %s: %s', $checkUrl, $response->getError()), $response->getReturnCode());
```

Comment:
The usual term is "Unexpected", we aren't Nostradamus :D
```suggestion
						throw new StorageException(sprintf('Unexpected error for %s: %s', $checkUrl, $response->getError()), $response->getReturnCode());
```